docusaurus/docs/how-to-guides/data-source-plugins/add-migration-handler-for-backend.md

```diff
@@ -0,0 +1,216 @@
+keywords:
+  - grafana
+  - plugins
+  - plugin
+  - migration
+  - datasource
+---
+
+As plugins evolve, significant changes may be required in the query model that the plugin defines. Reasons include major changes in the third-party service the plugin relies on, a refactor, new functionality, and more. When changing a data source query model, plugins should implement migration logic to transform existing queries from the previous format to the latest one. This migration logic should be included both in the backend (for non-frontend originated queries, such as alerts) and in the frontend (to adapt the query to newer versions of the `QueryEditor`).
+
+## Why add query migration handlers
+
+To ensure compatibility and maintain seamless performance, query migration handlers transform legacy queries into the current format. This approach allows you to deliver updates without breaking existing queries or duplicating code, offering users a smooth transition when they update your plugin.
+
+## Before you begin
+
+Depending on the approach you take for performing [one of the steps](#step-5-use-migration-code-from-the-frontend-using-experimental-apis) in this guide, you may need to fulfill certain prerequisites. These prerequisites are:
+
+1. Grafana must be configured to run data sources as standalone API servers, a behavior which is behind the feature flag [grafanaAPIServerWithExperimentalAPIs](https://github.com/grafana/grafana/blob/3457f219be1c8bce99f713d7a907ee339ef38229/pkg/services/featuremgmt/registry.go#L519).
+1. The plugin must be run on a Grafana version 11.4 or later.
+
+More information about these prerequisites is found in [step 5](#step-5-use-migration-code-from-the-frontend-using-experimental-apis) but if your plugin can't adhere to these requirements, there is an alternative approach using existing APIs
+
+## Implement a backend migration handler
+
+To implement a backend migration handler, add migration logic in your backend code. This guide walks you through the steps necessary to use the latest tooling to automatically migrate queries and avoid duplicating code.
+
+:::note
+
+The migration system detailed in this guide doesn't support two-way migrations. Only forward migrations are covered. Query migrations aren't automatically persisted, so users need to manually save changes to ensure the process works as expected.
+
+:::
+
+### Step 1 (optional): Add a query schema
+
+First of all, plugins don't need to have strongly typed queries. While this lowers the barrier for plugin development, plugins that don't define types are harder to scale and maintain. The first step in this guide is to add the required files to define the plugin query.
+
+See the following example: [grafana-plugin-examples#400](https://github.com/grafana/grafana-plugin-examples/pull/400). As you can see, there are multiple files to create. These files are used for both generating OpenAPI documentation and validating that the received queries are valid.
```

Comment:
> These files are used for both generating OpenAPI documentation

Can we link to something that explains where it can be used / found?

Reply:
Not really, it's explained below: "These schemas can be used for OpenAPI documentation, but it's a feature still in progress that isn't available yet". I can move that caveat here.